### DISCUSSION TITLE
Mark tests to expectedly fail

### DISCUSSION THREAD

**Discussion by andyjansson** [xunit/xunit#2936] [Ideas]
Test runners in other languages, such as pytest and jest, offer the ability to mark tests as expectedly failing (see pytest's `xfail` and jest's `failing` for reference). This means that when the test fails, it will be reported as such, but it will not produce an error code which would otherwise have broken the build. 

This feature is typically used when producing test cases for things which haven't been implemented yet or for bugs which are yet to be fixed. This is useful for things which aren't immediately fixable, but which you still need to track. 

It would be great to see such a feature in xunit, for instance by providing a flag in the attribute, e.g.:
```[Fact(Failing = true)]```

**Comment by bradwilson** [comment]
Is there value in actually running the test that's expected to fail? Today we offer `[Fact(Skip = "Reason")]` as the way to mark tests which are expected to fail (and therefore don't get run).

**Reply by andyjansson** [reply]
I think such a feature is valuable, hence why I am requesting it. It's also available in multiple other languages, so there is precedence. 

There's a few different reasons why I think these types of tests should be run. The reason why they're marked as failing is because they need to be fixed further down the line, primarily because the fix is non-trivial and requires the shaving of quite a few yaks. It might be that in the process of re-rearchitecting and fixing other bugs, that the circumstances for the failure might change and the test should be marked as passing. This is often the case with complex systems (for instance browsers, compilers, optimizers, etc.). This would have been invisible if the test was skipped. 

Failing tests also get more eyes on them, and there's an incentive to fix them. Skipped tests tend to get largely ignored, unfortunately. I think this is mainly due to psychology and assumptions. We (humans) tend to focus on things that are up-front, and we also assume that things which are skipped are skipped for a reason (e.g. being intended to be run in special environments or the like). 

When prompted to explain this feature, I like to call it a vastly superior TODO - One that doesn't go stale and doesn't require tons of investigation to orient yourself around because the specification is right in front of you.

There are other reasons which I haven't voiced, but I think this is sufficient to give you a gist of why I think they're useful and I hope I can convince you of their value as well. If you haven't already, I'd also like to request that you take a look at this feature as it exists in other languages to get a better understanding of its uses.

**Reply by bradwilson** [reply]
Let me see if I understand what you want when a test is marked as expecting failure:

- If it fails, still show the failure information but don't actually fail the test
- If it passes, then fail the test saying you expected it to fail

Does that sound right?

**Reply by andyjansson** [reply]
In the case that the test is marked with expecting failure, if it fails, it'll show the failure information but doesn't fail the test. If it passes, then it'll report the test as unexpectedly passing. 

In the latter case, jest opted to fail the test as you've described it, while pytest passes the test no matter what (see xfail and xpass). Personally, I'm fine with either.

**Reply by jalfd** [reply]
> Let me see if I understand what you want when a test is marked as expecting failure:
> 
>     * If it fails, still show the failure information but don't actually fail the test
> 
>     * If it passes, then fail the test saying you expected it to fail
> 
> 
> Does that sound right?

I agree, that would be useful.

Another use case where I'd find it helpful is as a first step in a PR.

1. Commit 1: introduce the test and show that it (expectedly) fails
1. Commit 2: Fix the issue and show that the test now passes

That tells a simpler story to the reviewer, and lets them easily verify that the test prior to the fix actually reproduces the issue.

**Reply by bradwilson** [reply]
To be clear:

> * If it fails, still show the failure information but don't actually fail the test

This is not possible with third party runners today like Test Explorer in Visual Studio. Collecting and displaying failure information _is only done when the test fails_.

**Reply by bradwilson** [reply]
The closest we could probably come here is to fabricate fake output information to attach to the test which describes how the test "failed" even though it passed (and of course, there's no way to differentiate these in the list from tests that are passing because they pass). What you're essentially asking for is a custom test result state, which third party runners don't really support.

**Comment by ojamte** [comment]
I have a test that fails because there are errors in the data that is used. Yes, it is an integration test.
I would like to be able to mark this test to be OK with a failure, but the output from the test should be shown. If the test pass, then all data is correct and I would like the test to pass as well. Since there is nothing that we are supposed to change in the code, I would not lika the failing test to fail the build. And I do not want the passing test to fail either.

it is a strech to use the builds and tests to show these errors or warnings, but it would be helpful. Looks lika a similar case as when you are creating the tests first for a rewrite as @andyjansson is mentioning. When the test passes, it should be OK. It looks like there is a need for yet another type, like "Ignore", that do run the test, but "IgnoreResult" or "IgnoreFailure", that makes those test visible with their results, if they do fail.

**Reply by bradwilson** [reply]
Well, this is the opposite of what OP wants. And frankly, I don't understand this idea of writing a test and then you just don't care whether it passes or fails. Whatever you're doing, it's not the kind of testing that we are aiming to support. I think it's probably best to not involve a testing framework at all; just write some code, run it, ignore the results, and move on with your day...? 🤷🏼‍♂️
